### PR TITLE
✨ Append original image's tags as new label

### PR DESCRIPTION
**Pull request by Cellane**
As discussed on Slack, this PR will add image’s original tags as an additional label.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -81,9 +82,14 @@ func main() {
 		os.Exit(1)
 	}
 
+	inspect, _, err := dockerClient.ImageInspectWithRaw(context.Background(), originalImageID.String())
```
Comment (josegonzalez) [inline main.go:85]: I didn't do this later, but we should handle errors in the same way we do above.

File: main.go [hunk 2]
```diff
@@ -81,9 +82,14 @@ func main() {
...
+	originalRepoTags, err := json.Marshal(inspect.RepoTags)
+	repoTagsLabel := fmt.Sprintf("com.dokku.docker-image-labeler/original-tags=%s", string(originalRepoTags))
+	appendLabels := append(*labels, repoTagsLabel)
```
Comment (josegonzalez) [inline main.go:88]: Probably don't need a new variable to hold `repoTagsLabel`.

File: main.go [hunk 3]
```diff
@@ -127,11 +133,6 @@ func main() {
 		os.Exit(1)
 	}
 
-	inspect, _, err := dockerClient.ImageInspectWithRaw(context.Background(), originalImageID.String())
-	if len(inspect.RepoTags) > 0 {
-		return
-	}
```
Comment (josegonzalez) [inline main.go:133]: We still want this return here, otherwise it'll delete all the other image tags that exist.

File: main.go [hunk 4]
```diff
@@ -81,9 +82,20 @@ func main() {
...
 	}
 
+	inspect, _, err := dockerClient.ImageInspectWithRaw(context.Background(), originalImageID.String())
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to inspect the source image (%s)\n", err.Error())
+	}
+
+	originalRepoTags, err := json.Marshal(inspect.RepoTags)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to encode image's original tags as JSON (%s)\n", err.Error())
```
Comment (josegonzalez) [inline main.go:92]: We also need to `os.Exit(1)` on error here and above.

Comment (Cellane) [reply main.go:92]: Ah, sorry!

**Comment by Cellane** [comment]
I think I might have found one additional issue, please do not merge yet.

**Comment by Cellane** [comment]
Dokku’s `ps:restart` triggers `docker-image-labeler` again, causing it to overwrite the already added `original-tags` label; so I changed this PR to only append the `original-tags` label if it is not already present.

**Comment by josegonzalez** [comment]
What was happening when it was "overwriting"? Can you show the label before/after with the previous version?

**Comment by Cellane** [comment]
Before last commit:

```shell
$ docker image tag haugene/transmission-openvpn:3.3 dokku/transmission:3.3
$ dokku tags:deploy transmission 3.3
$ docker container inspect transmission.web.1
```
```json
[
  {
    "…": "… omitted… ",
    "Config": {
      "…": "… omitted… ",
      "Labels": {
        "autoheal": "true",
        "com.dokku.app-name": "transmission",
        "com.dokku.container-type": "deploy",
        "com.dokku.docker-image-labeler/original-tags": "[\"dokku/transmission:3.3\",\"haugene/transmission-openvpn:3.3\"]",
        "com.dokku.dyno": "web.1",
        "com.dokku.image-stage": "release",
        "com.dokku.process-type": "web",
        "dokku": "",
        "org.label-schema.schema-version": "1.0",
        "org.label-schema.vendor": "dokku",
        "org.opencontainers.image.revision": "de2598df732811bfb4ac25395c5f27d6fa3b0d87",
        "org.opencontainers.image.source": "https://github.com/haugene/docker-transmission-openvpn"
      }
    }
  }
]
```
```shell
$ dokku ps:restart transmission
$ docker container inspect transmission.web.1
```
```json
[
  {
    "…": "… omitted… ",
    "Config": {
      "…": "… omitted… ",
      "Labels": {
        "autoheal": "true",
        "com.dokku.app-name": "transmission",
        "com.dokku.container-type": "deploy",
        "com.dokku.docker-image-labeler/original-tags": "[\"dokku/transmission:3.3\",\"dokku/transmission:latest\"]",
        "com.dokku.dyno": "web.1",
        "com.dokku.image-stage": "release",
        "com.dokku.process-type": "web",
        "dokku": "",
        "org.label-schema.schema-version": "1.0",
        "org.label-schema.vendor": "dokku",
        "org.opencontainers.image.revision": "de2598df732811bfb4ac25395c5f27d6fa3b0d87",
        "org.opencontainers.image.source": "https://github.com/haugene/docker-transmission-openvpn"
      }
    }
  }
]
```
```shell
$ docker image ls | grep transmission
haugene/transmission-openvpn                        3.3                          ff46391746b7   7 weeks ago     122MB
dokku/transmission                                  latest                       1635237376fb   41 years ago    122MB
dokku/transmission                                  3.3                          57cafa84f7cc   41 years ago    122MB
# Notice the different checksums for both images in the dokku/ namespace
```

After last commit:

```shell
$ docker image tag haugene/transmission-openvpn:3.3 dokku/transmission:3.3
$ dokku tags:deploy transmission 3.3
$ dokku ps:restart transmission
$ docker container inspect transmission.web.1
```
```json
  {
    "…": "… omitted… ",
    "Config": {
      "…": "… omitted… ",
      "Labels": {
        "autoheal": "true",
        "com.dokku.app-name": "transmission",
        "com.dokku.container-type": "deploy",
        "com.dokku.docker-image-labeler/original-tags": "[\"dokku/transmission:3.3\",\"haugene/transmission-openvpn:3.3\"]",
        "com.dokku.dyno": "web.1",
        "com.dokku.image-stage": "release",
        "com.dokku.process-type": "web",
        "dokku": "",
        "org.label-schema.schema-version": "1.0",
        "org.label-schema.vendor": "dokku",
        "org.opencontainers.image.revision": "de2598df732811bfb4ac25395c5f27d6fa3b0d87",
        "org.opencontainers.image.source": "https://github.com/haugene/docker-transmission-openvpn"
      }
    }
  }
]
```
```shell
$ docker image ls | grep transmission
haugene/transmission-openvpn                        3.3                          ff46391746b7   7 weeks ago     122MB
dokku/transmission                                  3.3                          57cafa84f7cc   41 years ago    122MB
dokku/transmission                                  latest                       57cafa84f7cc   41 years ago    122MB
# Checksum now remains the same
```

Basically what I think was happening is that Dokku ran `docker-image-labeler` during `ps:restart` on the `dokku/transmission:latest` image, which at that point was already once re-labelled by `docker-image-labeler`, thus did not contain in the `RepoTags` field what would be truly the original tags – they’d already be labelled in the `com.dokku.docker-image-labeler/original-tags` field, only to be promptly overwritten with invalid values.

**Comment by josegonzalez** [comment]
Should we extend the repo-tags list instead? As in, decode it if it exists and append any new repo tags to that list?

**Comment by Cellane** [comment]
Hmm… I wonder if that’s possible…
I have no idea how Docker works internally but I was somewhat under the impression that the `RepoTags` list must be dynamically computed based on what sort of tags exist on the system.

**Comment by josegonzalez** [comment]
I wonder if Dokku is like.. internally removing that tag on the image, even though we aren't using that tag?

**Comment by josegonzalez** [comment]
Merging so I can write tests and any refactors for next release.

**Pull request merged**